TestDerivs/src/test.cxx

```diff
@@ -110,64 +182,170 @@ extern "C" void TestDerivs_CalcDerivs(CCTK_ARGUMENTS) {
+              Derivs::calc_deriv_upwind<8>(gf2_chi, mask, p.I, dx,
+                                           gf_beta(mask, index2)));
+        });
+    break;
+  }
+  default:
+    assert(0);
+  }
 
 #if CCTK_DEBUG
   grid.loop_int_device<0, 0, 0>(
       grid.nghostzones, [=] ARITH_DEVICE(const PointDesc &p) ARITH_INLINE {
-        using std::cos, std::sin;
-        const auto sinx = sin(p.x);
-        const auto siny = sin(p.y);
-        const auto sinz = sin(p.z);
-        const auto cosx = cos(p.x);
-        const auto cosy = cos(p.y);
-        const auto cosz = cos(p.z);
-        const auto dxxdchi =
-            -2 * kxx * cosx * cosx + 2 * kxx * sinx * sinx - kxy * cosx * siny;
-        const auto dxydchi = -kxy * cosy * sinx;
-        const auto dxzdchi = 0.0;
-        const auto dyydchi = -kxy * cosx * siny - kyz * siny * sinz;
-        const auto dyzdchi = kyz * cosy * cosz;
-        const auto dzzdchi = -kyz * siny * sinz;
+        Arith::vec<CCTK_REAL, dim> dchi;
+        Arith::smat<CCTK_REAL, dim> ddchi;
+        poly_derivs(kxx, kxy, kyz, p.x, p.y, p.z, dchi, ddchi);
 
         const auto tiny = 5e-3;
-        if (dxxdchi - gf_ddchi(0, 0)(p.I) > tiny ||
-            dxydchi - gf_ddchi(0, 1)(p.I) > tiny ||
-            dxzdchi - gf_ddchi(0, 2)(p.I) > tiny ||
-            dyydchi - gf_ddchi(1, 1)(p.I) > tiny ||
-            dyzdchi - gf_ddchi(1, 2)(p.I) > tiny ||
-            dzzdchi - gf_ddchi(2, 2)(p.I) > tiny) {
+        if (dchi(0) - gf_dchi(0)(p.I) > tiny ||
+            dchi(1) - gf_dchi(1)(p.I) > tiny ||
+            dchi(2) - gf_dchi(2)(p.I) > tiny ||
+            ddchi(0, 0) - gf_ddchi(0, 0)(p.I) > tiny ||
+            ddchi(0, 1) - gf_ddchi(0, 1)(p.I) > tiny ||
+            ddchi(0, 2) - gf_ddchi(0, 2)(p.I) > tiny ||
+            ddchi(1, 1) - gf_ddchi(1, 1)(p.I) > tiny ||
+            ddchi(1, 2) - gf_ddchi(1, 2)(p.I) > tiny ||
+            ddchi(2, 2) - gf_ddchi(2, 2)(p.I) > tiny) {
 #ifndef SYCL_LANGUAGE_VERSION
-          printf("ddxx = %f\n", dxxdchi - gf_ddchi(0, 0)(p.I));
-          printf("ddxy = %f\n", dxydchi - gf_ddchi(0, 1)(p.I));
-          printf("ddxz = %f\n", dxzdchi - gf_ddchi(0, 2)(p.I));
-          printf("ddyy = %f\n", dyydchi - gf_ddchi(1, 1)(p.I));
-          printf("ddyz = %f\n", dyzdchi - gf_ddchi(1, 2)(p.I));
-          printf("ddzz = %f\n", dzzdchi - gf_ddchi(2, 2)(p.I));
+          printf("dx = %f\n", dchi(0) - gf_dchi(0)(p.I));
+          printf("dy = %f\n", dchi(1) - gf_dchi(1)(p.I));
+          printf("dz = %f\n", dchi(2) - gf_dchi(2)(p.I));
+          printf("ddxx = %f\n", ddchi(0, 0) - gf_ddchi(0, 0)(p.I));
+          printf("ddxy = %f\n", ddchi(0, 1) - gf_ddchi(0, 1)(p.I));
+          printf("ddxz = %f\n", ddchi(0, 2) - gf_ddchi(0, 2)(p.I));
+          printf("ddyy = %f\n", ddchi(1, 1) - gf_ddchi(1, 1)(p.I));
+          printf("ddyz = %f\n", ddchi(1, 2) - gf_ddchi(1, 2)(p.I));
+          printf("ddzz = %f\n", ddchi(2, 2) - gf_ddchi(2, 2)(p.I));
 #endif
           assert(0);
```

Comment:
don't use `assert(0)` if it's an error use `CCTK_VERROR` or `abort()`. `assert` is turned into a no-op by `NDEBUG`.